neofs_contract.go

```diff
@@ -405,3 +422,61 @@ func verifySignatures(irList []node, data []byte, message []byte, offset int) bo
 	runtime.Log("not enough verified signatures")
 	return false
 }
+
+// isInnerRingRequest returns true if contract was invoked by inner ring node.
+func isInnerRingRequest(irList []node) bool {
+	for i := 0; i < len(irList); i++ {
+		irNode := irList[i]
+
+		if runtime.CheckWitness(irNode.pub) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// todo: votes must be from unique inner ring nods
+func vote(ctx storage.Context, id []byte) int {
+	var (
+		newCandidates = []ballot{}
+		candidates    = getSerialized(ctx, voteKey).([]ballot)
+		found         = -1
+	)
+
+	for i := 0; i < len(candidates); i++ {
+		cnd := candidates[i]
+		if util.Equals(cnd.id, id) {
+			cnd = ballot{id: id, n: cnd.n + 1}
+			found = cnd.n
+		}
+		newCandidates = append(newCandidates, cnd)
+	}
+
+	if found < 0 {
+		newCandidates = append(newCandidates, ballot{id: id, n: 1})
+		found = 1
+	}
+
+	data := runtime.Serialize(newCandidates)
+	storage.Put(ctx, voteKey, data)
+
+	return found
+}
+
+func removeVotes(ctx storage.Context, id []byte) {
+	var (
+		newCandidates = []ballot{}
+		candidates    = getSerialized(ctx, voteKey).([]ballot)
```

Comment:
Can we just cut an element from `candidates` slice below instead of filling `newCandidates` slice?

Reply:
No, for now we can't: https://github.com/nspcc-dev/neo-go/issues/999